File: expectanim/src/main/java/com/github/florent37/expectanim/ViewCalculator.java
package com.github.florent37.expectanim;

import android.view.View;

import java.util.HashMap;
import java.util.Map;

import static java.lang.Math.abs;
import static java.lang.Math.cos;
import static java.lang.Math.sin;

/**
 * Created by florentchampigny on 20/02/2017.
 */

public class ViewCalculator {
    private final Map<View, ViewExpectation> expectationForView;

    private boolean enableRotation = true;
    private boolean enableScale = true;

    public ViewCalculator() {
        expectationForView = new HashMap<>();
    }

    public void setExpectationForView(View view, ViewExpectation viewExpectation) {
        expectationForView.put(view, viewExpectation);
    }

    public void wasCalculated(ViewExpectation viewExpectation) {
        //no-op
    }

    public float finalPositionLeftOfView(View view, boolean itsMe) {
        Float finalX = null;

        final ViewExpectation viewExpectation = expectationForView.get(view);
        if (viewExpectation != null) {
            final Float futurPositionX = viewExpectation.getFuturPositionX();
            if (futurPositionX != null) {
                finalX = futurPositionX;
            }
        }

        if (finalX == null) {
            finalX = view.getX();
        }

        if (itsMe) {
            finalX -= (view.getWidth() - this.finalWidthOfView(view)) / 2f;
        }

        return finalX;
    }

    public float finalPositionLeftOfView(View view) {
        return finalPositionLeftOfView(view, false);
    }

    public float finalPositionRightOfView(View view) {
        return finalPositionLeftOfView(view) + finalWidthOfView(view);
    }

    public float finalPositionTopOfView(View view, boolean itsMe) {
        Float finalTop = null;

        final ViewExpectation viewExpectation = expectationForView.get(view);
        if (viewExpectation != null) {
            final Float futurPositionY = viewExpectation.getFuturPositionY();
            if (futurPositionY != null) {
                finalTop = futurPositionY;
            }
        }

        if (finalTop == null) {
            finalTop = 1f * view.getTop();
        }

        if (itsMe) {
            finalTop -= (view.getHeight() - this.finalHeightOfView(view)) / 2f;
        }

        return finalTop;
    }

    public float finalPositionTopOfView(View view) {
        return finalPositionTopOfView(view, false);
    }

    public float finalPositionBottomOfView(View view) {
        return finalPositionTopOfView(view) + finalHeightOfView(view);
    }

    public float finalCenterXOfView(View view) {
        if (expectationForView.containsKey(view)) {
            return expectationForView.get(view).getFuturPositionX() + finalWidthOfView(view) / 2f;
        } else {
            return view.getLeft() + view.getWidth() / 2f;
        }
    }

    public float finalCenterYOfView(View view) {
        if (expectationForView.containsKey(view)) {
            return expectationForView.get(view).getFuturPositionY() + finalHeightOfView(view) / 2f;
        } else {
            return view.getTop() + view.getHeight() / 2f;
        }
    }

    public float finalWidthOfView(View view) {
        float with = view.getWidth();
        if (expectationForView.containsKey(view)) {
            final ViewExpectation viewExpectation = expectationForView.get(view);

            with = widthScaled(viewExpectation, view, with);

            if (enableRotation) {
                with = widthRotated(viewExpectation, view, with);
            }
        }
        return with;
    }

    private float widthScaled(ViewExpectation viewExpectation, View view, float width) {
        final Float scaleX = viewExpectation.getWillHasScaleX();
        if (scaleX != 1f) {
            width = scaleX * width; // + view.getPivotX() * scaleX;
        }
        return width;
    }

    private float widthRotated(ViewExpectation viewExpectation, View view, float width) {
        final Float willHaveRotationX = viewExpectation.getWillHaveRotation();
        if (willHaveRotationX != null) {
            final double radians = Math.toRadians(90 - willHaveRotationX);
            width = (float) (abs(width * sin(radians)) + abs(heightScaled(viewExpectation, view, view.getHeight()) * cos(radians)));
        }
        return width;
    }

    public float finalHeightOfView(View view) {
        float height = view.getHeight();
        if (expectationForView.containsKey(view)) {
            final ViewExpectation viewExpectation = expectationForView.get(view);

            height = heightScaled(viewExpectation, view, height);

            if (enableRotation) {
                height = heightRotated(viewExpectation, view, height);
            }
        }
        return height;
    }

    private float heightScaled(ViewExpectation viewExpectation, View view, float height) {
        final Float scaleY = viewExpectation.getWillHasScaleY();
        if (scaleY != 1) {
            height = scaleY * height; // + view.getPivotY() * scaleY;
        }
        return height;
    }

    private float heightRotated(ViewExpectation viewExpectation, View view, float height) {
        final Float willHaveRotationX = viewExpectation.getWillHaveRotation();
        if (willHaveRotationX != null) {
            final double radians = Math.toRadians(willHaveRotationX);
            height = (float) (abs(height * cos(radians)) + abs(widthScaled(viewExpectation, view, view.getWidth()) * sin(radians)));
        }
        return height;
    }


}

File: kotlinpleaseanimate/src/main/java/com/github/florent37/kotlin/pleaseanimate/ViewCalculator.kt
package com.github.florent37.kotlin.pleaseanimate

import android.view.View

import java.lang.Math.abs
import java.lang.Math.cos
import java.lang.Math.sin

class ViewCalculator {
    private val expectationForView: MutableMap<View, ViewExpectation> = mutableMapOf()

    private val enableRotation = true
    private val enableScale = true

    fun setExpectationForView(view: View, viewExpectation: ViewExpectation) {
        expectationForView.put(view, viewExpectation)
    }

    fun wasCalculated(viewExpectation: ViewExpectation) {
        //no-op
    }

    @JvmOverloads
    fun finalPositionLeftOfView(view: View, itsMe: Boolean = false): Float {
        var finalX: Float? = null

        val viewExpectation = expectationForView[view]
        if (viewExpectation != null) {
            val futurPositionX = viewExpectation.getFuturPositionX()
            if (futurPositionX != null) {
                finalX = futurPositionX
            }
        }

        if (finalX == null) {
            finalX = view.x
        }

        if (itsMe) {
            finalX -= (view.width - this.finalWidthOfView(view)) / 2f
        }

        return finalX
    }

    fun finalPositionRightOfView(view: View): Float {
        return finalPositionLeftOfView(view) + finalWidthOfView(view)
    }

    @JvmOverloads
    fun finalPositionTopOfView(view: View, itsMe: Boolean = false): Float {
        var finalTop: Float? = null

        val viewExpectation = expectationForView[view]
        if (viewExpectation != null) {
            val futurPositionY = viewExpectation.getFuturPositionY()
            if (futurPositionY != null) {
                finalTop = futurPositionY
            }
        }

        if (finalTop == null) {
            finalTop = 1f * view.top
        }

        if (itsMe) {
            finalTop -= (view.height - this.finalHeightOfView(view)) / 2f
        }

        return finalTop
    }

    fun finalPositionBottomOfView(view: View): Float {
        return finalPositionTopOfView(view) + finalHeightOfView(view)
    }

    fun finalCenterXOfView(view: View): Float {
        return if (expectationForView.containsKey(view)) {
            expectationForView[view]!!.getFuturPositionX()!! + finalWidthOfView(view) / 2f
        } else {
            view.left + view.width / 2f
        }
    }

    fun finalCenterYOfView(view: View): Float {
        return if (expectationForView.containsKey(view)) {
            expectationForView[view]!!.getFuturPositionY()!! + finalHeightOfView(view) / 2f
        } else {
            view.top + view.height / 2f
        }
    }

    fun finalWidthOfView(view: View): Float {
        var with = view.width.toFloat()
        if (expectationForView.containsKey(view)) {
            val viewExpectation = expectationForView[view]

            with = widthScaled(viewExpectation!!, view, with)

            if (enableRotation) {
                with = widthRotated(viewExpectation, view, with)
            }
        }
        return with
    }

    private fun widthScaled(viewExpectation: ViewExpectation, view: View, width: Float): Float {
        var w = width
        val scaleX = viewExpectation.getWillHasScaleX()
        if (scaleX != 1f) {
            w = scaleX * width // + view.getPivotX() * scaleX;
        }
        return w
    }

    private fun widthRotated(viewExpectation: ViewExpectation, view: View, width: Float): Float {
        var w = width
        val willHaveRotationX = viewExpectation.getWillHaveRotation()
        if (willHaveRotationX != null) {
            val radians = Math.toRadians((90f - willHaveRotationX).toDouble())
            w = (abs(width * sin(radians)) + abs(heightScaled(viewExpectation, view, view.height.toFloat()) * cos(radians))).toFloat()
        }
        return w
    }

    fun finalHeightOfView(view: View): Float {
        var height = view.height.toFloat()
        if (expectationForView.containsKey(view)) {
            val viewExpectation = expectationForView[view]

            height = heightScaled(viewExpectation!!, view, height)

            if (enableRotation) {
                height = heightRotated(viewExpectation, view, height)
            }
        }
        return height
    }

    private fun heightScaled(viewExpectation: ViewExpectation, view: View, height: Float): Float {
        var h = height
        val scaleY = viewExpectation.getWillHasScaleY()
        if (scaleY != 1f) {
            h = scaleY * height // + view.getPivotY() * scaleY;
        }
        return h
    }

    private fun heightRotated(viewExpectation: ViewExpectation, view: View, height: Float): Float {
        var h = height
        val willHaveRotationX = viewExpectation.getWillHaveRotation()
        if (willHaveRotationX != null) {
            val radians = Math.toRadians(willHaveRotationX.toDouble())
            h = (abs(height * cos(radians)) + abs(widthScaled(viewExpectation, view, view.width.toFloat()) * sin(radians))).toFloat()
        }
        return h
    }


}
